enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,605 @@
+
+Initial proofs-of-concept:
+
+- [cluster-monitoring-operator #1590](https://openshift/cluster-monitoring-operator/pull/1590)
+
+  This is the initial PR adding the `AlertingRule` controller to c-m-o.
+
+- [cluster-monitoring-operator #1596](https://github.com/openshift/cluster-monitoring-operator/pull/1596)
+
+  This is the initial PR adding the `AlertRelabelConfig` controller to c-m-o.
+
+- [prometheus-operator #4609](https://github.com/prometheus-operator/prometheus-operator/pull/4609)
+
+  This is an attempt to introduce the `AlertRelabelConfig` type in upstream
+  prometheus-operator.  If this isn't accepted upstream, it should be
+  straightforward to bring downstream into c-m-o.
+
+  *Update*: Looks like this isn't a fit for upstream.  It has been moved to
+  cluster-monitoring-operator.
+
+## Drawbacks
+
+TBD
+
+## Alternatives
+
+- Primary alternative: https://github.com/openshift/enhancements/pull/958
+
+  This is an alternative to the original alert overrides proposal, which
+  involves a new CRD which matches existing alerting rules by name and labels,
+  and then allows overriding any field.  The original is more flexible, but
+  significantly more complex, and presents UX/UI challenges for console.
+
+  The original proposal aims to allow overriding arbitrary fields in existing
+  alerting rules. To do that, a new resource type takes a list of alert
+  definitions that each have a selector to target an existing alert. That turns
+  out to be complicated because the alerts are spread across a bunch of
+  PrometheusRule resources, the names don't have to be unique, and outside of
+  name we can only match on static labels, which aren't used very heavily except
+  for setting a severity.
+
+  This introduces a bunch of edge cases: What happens when multiple alerts
+  match? Currently an error condition. What happens when the match used to be
+  exact, but a new alert is introduced that also matches? Also an error
+  condition, but one the user isn't likely to notice. What happens if the
+  targeted alert is removed entirely? The patched alert also disappears. This
+  all works technically, but the UX feels a bit odd in my opinion.
+
+  There are also issues with integrating with the console. The overridden
+  alerting rules and alerts still show up in the Prometheus API. The console
+  should filter these somehow, but the original alerts must remain unchanged, so
+  it now has to filter itself based on the original alerts and information from
+  the overrides.
+
+- Another alternative is to have a namespace, e.g. `openshift-user-alerts`,
+  dedicated for user configuration objects.  Users can simply create any custom
+  `PrometheusRule` objects in this namespace, and the operator can ensure they
+  are then picked up by platform monitoring and labeled as user-defined.
+
+  The primary disadvantage to this is that we have less control.  We can't for
+  example disallow creation of recording rules with this method.
```

Comment:
There's been a lot of discussion on the new `AlertingRule` type, which is understandable. It's a fairly complex type, that's actually just a copy of upstream, that OpenShift will have to support forever. Given that, I thought it might make sense to revisit this briefly.

I think we *could* throw out the `AlertingRule` type but keep the simpler `AlertRelabelConfig` bits, then have a new key in the `cluster-monitoring-config` `ConfigMap` along the lines of `enableUserDefinedAlerting` -- when that is enabled, cluster-monitoring-operator could create the `openshift-user-alerts` namespace and deploy admission webhooks for only the `PrometheusRule` objects in this new namespace. These webhooks could do the work of disallowing recording rules or whatever other policy should be enforced. I think that could be a way to accomplish the same as what we have here, but with less new API surface in OpenShift.

cc: @jan--f @simonpasquier

Reply:
Thanks for outlining another path.
I think at this moment I would still favor the API route. This is arguably the right place to specify that.
The other option you outlined (lets call it the admission webhook route) does provide a slightly worse user experience. They can create a `PrometheusRule` object but it gets rejected. Granted that feedback loop is very short with the admission webhook, but friction still exists. Additionally I suspect we will get queries why we decide to reject those with the respective RFEs. An API definition sets an expectation if you will.
On the other hand I think the risk of adding the new API type is fairly manageable. If the `PrometheusRule` object ever changes, we certainly have to reflect that and write code to translate between versions. Though I would expect that a) dramatic changed to `PrometheusRule` are unlikely and b) if a is incorrect we'd be able to reuse a lot of the respective upstream compatibility code.
Iiuc the complexity of the new type stems from the intention to keep this as close as possible to the "parent" type you are deriving this from. This should make any future adjustments as easy as possible (reuse compatibility code from the "parent"), so I think there is an argument to be made for making this addition as complex as it is.